File: GOLANG/InterfacesProtocolos.go
package main

import "fmt"

type Imprimivel interface {
    Imprimir() string
}

type Relatorio struct {
    Conteudo string
}

func (r Relatorio) Imprimir() string {
    return fmt.Sprintf("Imprimindo relatório: %s", r.Conteudo)
}

type Contrato struct {
    Partes []string
}

func (c Contrato) Imprimir() string {
    return fmt.Sprintf("Imprimindo contrato entre: %s", c.Partes)
}

func imprimirDocumentos(documentos []Imprimivel) {
    for _, doc := range documentos {
        fmt.Println(doc.Imprimir())
    }
}

func main() {
    relatorio := Relatorio{Conteudo: "Dados financeiros do Q1"}
    contrato := Contrato{Partes: []string{"Empresa A", "Empresa B"}}

    documentos := []Imprimivel{relatorio, contrato}

    imprimirDocumentos(documentos)
}

File: JAVA/InterfacesProtocolos.java
interface Imprimivel {
    String imprimir();
}

class Relatorio implements Imprimivel {
    private String conteudo;

    public Relatorio(String conteudo) {
        this.conteudo = conteudo;
    }

    @Override
    public String imprimir() {
        return "Imprimindo relatório: " + conteudo;
    }
}

class Contrato implements Imprimivel {
    private String partes;

    public Contrato(String partes) {
        this.partes = partes;
    }

    @Override
    public String imprimir() {
        return "Imprimindo contrato entre: " + partes;
    }
}

public class Main {
    public static void main(String[] args) {
        Imprimivel relatorio = new Relatorio("Dados financeiros do Q1");
        Imprimivel contrato = new Contrato("Empresa A e Empresa B");

        System.out.println(relatorio.imprimir());
        System.out.println(contrato.imprimir());
    }
}
